src/partition/repart.c

```diff
@@ -6963,6 +6961,38 @@ static int parse_argv(int argc, char *argv[]) {
                         *p = gpt_partition_type_override_architecture(*p, arg_architecture);
         }
 
+        if (arg_signing_provider && !private_key)
+                return log_error_errno(
+                                SYNTHETIC_ERRNO(EINVAL),
+                                "--private-key= must be specified when --signing-provider= is used.");
+
+        if (private_key) {
+                if (isempty(arg_signing_provider)) {
+                        _cleanup_(erase_and_freep) char *k = NULL;
+                        size_t n = 0;
+
+                        r = read_full_file_full(
+                                        AT_FDCWD, private_key, UINT64_MAX, SIZE_MAX,
+                                        READ_FULL_FILE_SECURE|READ_FULL_FILE_WARN_WORLD_READABLE|READ_FULL_FILE_CONNECT_SOCKET,
+                                        NULL,
+                                        &k, &n);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to read key file '%s': %m", private_key);
+
+                        r = parse_private_key(k, n, &arg_private_key);
+                        if (r < 0)
+                                return r;
```

Comment:
Can't we always go through the provider codepath? It should support regular paths as uris just fine no?

Reply:
Yes you can, the store api is fully generic and will load pem files just fine, you do not need special code if you use modern openssl 3 store apis, and that is the recommended way to deal with the pkcs11-provider, rather than having custom code.
Just rely on interested admins (or the distributions to do it for them) to configure the provider in the openssl.cnf and just provide a pkcs11 uri instead of a file name for the keys and all should just work transparently.

Reply:
And what about --signing-provider in that case? Does it even make sense to have such an argument? Or is it sufficient to start interpreting the --private-key= argument which currently just takes a file path as a URI. Then if we need to fallback to the engine stuff maybe we can determine the engine to use from the uri as well?

Reply:
> Can't we always go through the provider codepath? It should support regular paths as uris just fine no?

eh not really, given providers don't really work at the moment

Reply:
> eh not really, given providers don't really work at the moment

Not with your token, but with files on disk without pins they should work fine I think

Reply:
I think I've done enough experimenting for a while, I just want to be able to use my yubikey

Reply:
@DaanDeMeyer as I said the default path without any HW toke invovled can be changed to use the Store api by default. And then everything will work automatically as the store api select the correct provider based on the URI it gets provived with, and fo course if a URI is not handled by any provider then it will throw an error.
going that way you will be able to use also other providers without code changes like the tpm2 provider or the oqs-provider etc...

Reply:
> Can't we always go through the provider codepath? It should support regular paths as uris just fine no?

Our loader is a bit fancy though, i.e. it will warn about world-readable access modes, and will connect to AF_UNIX sockets if you reference one. We try to make sure all our tools that read keys work that way. I doubt that openssl supports that (or wants to support that)